ta/pkcs11/src/processing_symm.c

```diff
@@ -751,6 +857,73 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 
 		break;
 
+	case PKCS11_CKM_MD5_HMAC_GENERAL:
+	case PKCS11_CKM_SHA_1_HMAC_GENERAL:
+	case PKCS11_CKM_SHA224_HMAC_GENERAL:
+	case PKCS11_CKM_SHA256_HMAC_GENERAL:
+	case PKCS11_CKM_SHA384_HMAC_GENERAL:
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		assert(session->processing->extra_ctx);
+		hmac_len = *(uint32_t *)session->processing->extra_ctx;
+
+		switch (function) {
+		case PKCS11_FUNCTION_SIGN:
+			if (out_size < hmac_len) {
+				/* inform client of required size */
+				out_size = hmac_len;
+				output_data = true;
+				rc = PKCS11_CKR_BUFFER_TOO_SMALL;
+				goto out;
+			}
+
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size,
+						  computed_mac,
+						  &computed_mac_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
+				memcpy(out_buf, computed_mac, hmac_len);
+			}
+			if (res == TEE_SUCCESS || res == TEE_ERROR_SHORT_BUFFER)
+				output_data = true;
+
+			/* inform client of required size */
+			out_size = hmac_len;
+			rc = tee2pkcs_error(res);
+			break;
+		case PKCS11_FUNCTION_VERIFY:
+			/* must compute full MAC before comparing partial */
+			res = TEE_MACComputeFinal(proc->tee_op_handle, in_buf,
+						  in_size, computed_mac,
+						  &computed_mac_size);
+
+			if (!in2_size || in2_size > computed_mac_size) {
+				EMSG("Invalid signature size: %"PRIu32,
+				     in2_size);
+				return PKCS11_CKR_SIGNATURE_LEN_RANGE;
+			}
+
+			if (res == TEE_SUCCESS) {
+				/*
+				 * Only the first in2_size bytes of the
+				 * signature to be verified is passed in from
+				 * NW
```

Comment:
NW -- perhaps REE (or be more verbose) as that is commonly used with TEE terminology.

Or just passed from "caller" then if someone does TA->TA call then it would also be correct.

Reply:
Agree with "caller" or even "client".

Reply:
Changed to "caller"